index.d.ts

```diff
@@ -0,0 +1,200 @@
+	TCbArgs = Parameters<Exclude<TCb, undefined>>
+> =
+	TCbArgs extends [CallbackError?] ? void :
+	TCbArgs extends [CallbackError, infer U] ? U :
+	TCbArgs extends any[] ? TCbArgs[1] extends undefined ? void : TCbArgs[1]:
+	never;
+
+type CallbackMultiReturnType<
+	T extends AnyFunction,
+	TCb extends AnyFunction = LastParameter<T>,
+	TCbArgs = Parameters<Exclude<TCb, undefined>>
+> =
+	TCbArgs extends [any, ... infer U] ? U :
+	TCbArgs extends [any?, ... infer U] ? U :
+	TCbArgs extends [any] ? [] :
+	TCbArgs extends [any?] ? [] :
+	TCbArgs extends [] ? [] :
+	never;
+
+type CallbackWithoutErrorSingleReturnType<T extends AnyFunction> = Parameters<LastParameter<T>> extends [] ? void : Parameters<LastParameter<T>>[0];
+
+type CallbackWithoutErrorMultiReturnType<T extends AnyFunction> = Parameters<LastParameter<T>>;
+
+type CallbackReturn<T extends AnyFunction, MultiArgs extends boolean, ErrorFirst extends boolean> = SelectByOptionOr<ErrorFirst,
+	SelectByOptionOr<MultiArgs, CallbackMultiReturnType<T>, CallbackSingleReturnType<T>>,
+	SelectByOptionOr<MultiArgs, CallbackWithoutErrorMultiReturnType<T>, CallbackWithoutErrorSingleReturnType<T>>>;
+
+type PifiedFunction<F extends AnyFunction, O extends PifyOptions> = Resolve<(...args:ParametersExceptLast<F>) => Promise<CallbackReturn<F, GetMultiArgs<O>, GetErrorFirst<O>>>>;
+
+type GetOption<O extends PifyOptions, Name extends string, Default extends boolean> =
+	O extends {[F in Name]: true} ? true :
+	O extends {[F in Name]: false} ? false :
+	O extends {[F in Name]: boolean} ? boolean :
+	O extends {[F in Name]: undefined} ? Default :
+	Default;
+type GetExcludeMain<O extends PifyOptions> = Resolve<GetOption<O, 'excludeMain', false>>;
+type GetErrorFirst<O extends PifyOptions> = Resolve<GetOption<O, 'errorFirst', true>>;
+type GetMultiArgs<O extends PifyOptions> = Resolve<GetOption<O, 'multiArgs', false>>;
+
+type PifiedObject<T, O extends PifyOptions> = Resolve<{
+	[Key in (keyof T)]: T[Key] extends AnyFunction ? PifiedFunction<T[Key], O> : T[Key];
+}>;
+
+type PifiedObjectFunction<T, O extends PifyOptions> = Resolve<
+	GetExcludeMain<O> extends true ?
+		T extends AnyFunction	?
+			({} extends T ? T & PifiedObject<T, O> : T) :
+			PifiedObject<T, O> :
+		T extends AnyFunction ?
+			LastParameter<T> extends never ? never : LastParameter<T> extends AnyFunction ?
+				({} extends T ? PifiedFunction<T, O> & PifiedObject<T, O> : PifiedFunction<T, O>) :
+				never :
+			PifiedObject<T, O>
+	>;
+
+interface PifyOptions {
+	/**
+	By default, the promisified function will only return the second argument from the callback, which works fine for most APIs. This option can be useful for modules like `request` that return multiple arguments. Turning this on will make it return an array of all arguments from the callback, excluding the error argument, instead of just the second argument. This also applies to rejections, where it returns an array of all the callback arguments, including the error.
```

Comment:
```suggestion
	By default, the promisified function will only return the second argument from the callback, which works fine for most APIs. This option can be useful for modules like `request` that return multiple arguments. Turning this on will make it return an array of all arguments from the callback, excluding the error argument, instead of just the second argument. This also applies to rejections, where it returns an array of all the callback arguments, including the error.

```